### ISSUE TITLE
Zot extensions 

### ISSUE THREAD

**Issue by rchamarthy** [project-zot/zot#50]
# Goal

Zot should provide an extension mechanism to implement additional features outside the OCI specification. Extensions provides a way to implement experimental/draft features to the OCI community and gain experience with these use cases. If and when these features are standardized, zot will promote the extensions to core feature set and can be the reference implementation of the specification.

All extensions are compiled in to the zot binary and can be selectively enabled/disabled using runtime configuration.

## Extensions

Zot supports two types of extensions. Inline extensions which are called as requests are being processed by zot. Asynchronous extensions which are called after an event finished.

### Inline Extension

Extension points that are called inline with a repository feature like push image, pull image and list, create, update, delete manifests and all operations. We will use the middleware pattern to register the extension points that will extend each of this feature and can react to these events inline as they are being processed by zot.

### Async Extension

Async extensions are invoked in their own context in reaction to an event. For example CRUD operations on an image or a repository. There is no ordering guarantee for the extensions, all of them may be invoked in parallel. 

## Configuration

Each extension must be explicitly configured and only configured extensions are registered with runtime and invoked appropriately. Here is a sample configuration

```json
{
  "extensions": [
    {"scan": { "scanInterval" : "10m" }},
    {"query": {"queryPort" : 6000}},
  ]
}
```

**Comment by jdolitsky** [comment]
Was about to open a new issue related, but came across this... Is there anyway that this can implement what is being proposed in https://github.com/opencontainers/distribution-spec/discussions/238 ?

Could this be expanded to enable additional HTTP endpoints? cc @rchincha

**Comment by jdolitsky** [comment]
> All extensions are compiled in to the zot binary

What if this were not a requirement, and the extension was developed out of tree, in a different language, etc.?

Project Octant has an interesting approach to plugin development: https://reference.octant.dev/?path=/docs/docs-plugins-1-getting-started--page

**Comment by rchincha** [comment]
> 
> 
> Was about to open a new issue related, but came across this... Is there anyway that this can implement what is being proposed in [opencontainers/distribution-spec#238](https://github.com/opencontainers/distribution-spec/discussions/238) ?
> 
> Could this be expanded to enable additional HTTP endpoints? cc @rchincha

@jdolitsky we have already organized our code as "core" dist-spec end-points and additional functionality as "extensions" [1].
Some of our extensions show up as graphQL endpoints for faster searches, vulnerability scanning etc. The separation is enforced at build time as well, you can have a dist-spec-only zot-minimal [2] or the whole thing [3].

We are definitely looking at notary-v2 (@SteveLastker) as one of the next extensions to add.

[1] https://github.com/anuvu/zot/tree/master/pkg/extensions
[2] https://github.com/anuvu/zot/blob/master/Makefile#L14
[3] https://github.com/anuvu/zot/blob/master/Makefile#L18

**Comment by rchincha** [comment]
> 
> 
> > All extensions are compiled in to the zot binary
> 
> What if this were not a requirement, and the extension was developed out of tree, in a different language, etc.?
> 
> Project Octant has an interesting approach to plugin development: https://reference.octant.dev/?path=/docs/docs-plugins-1-getting-started--page

So far this is not an approach we have taken. But if there is a compelling plugin/functionality that this project can benefit from, will absolutely consider it. Perhaps any additional endpoints (to expose said plugin functionality) are also plugged in dynamically.

**Comment by mikemccracken** [comment]
@rchamarthy is this moot with the current extension support?
